# **Elements of Computing Final Exam_Part 1:** Luxury Dealership Business Use Case using SQLite with Python


---

# **Project Overview**

As the in-house Marketing and Information Specialist for Elite Motor Dealership, you will analyze historical sales data from the `Luxury_Car_Dlrshp.db` database to drive data-led decision-making at the dealership.

# **Your Task:**

* **Step One -** Upload database to notebook environment. (0 pts)
* **Step Two -** Imports and SQLite database connection. (0 pts)
* **Step Three -** Confirm all tables present in database (0 pts)
* **Step Four -** Complete correctly the (5) SQL queries for total revenue and profit KPIs prompts. (5 pts / each)
* **Step Five -** Complete correctly the (3) SQL queries for customer and sales performance prompts. (5 pts / each)
* **Step Six -** Complete correctly the (2) SQL queries for trade-ins and finance analysis prompts. (5 pts / each)

###**For each prompt, you must:**

- **Task 1:** Execute the correct SQL code.

> **Note** - *"correct"*   SQL code is a query that returns an output that completes all elements of the original prompt.

- **Task 2:** Provide a clear answer to the prompt provided for each query based on your returned output.

- **Task 3:** Offer a strategic suggestion for management at Elite Motor Dealership, identifying how these insights (query returns) can improve business KPIs and or reveal new market opportunities.


# **STEP ONE**

*   Upload the `Luxury_Car_Dlrshp.db` into the "Files" section of this notebook.

    **Note** - This file should appear to the left of this cell once successfully added to "Files".

# **STEP TWO**

*   Import all necessary libraries and packages.
*   Establish the path to the DB stored in Files.

In [1]:
import sqlite3
import pandas as pd

# Set the path to the SQLite database
db_path = "./Luxury_Car_Dlrshp.db"

# Connect to SQLite database
conn = sqlite3.connect(db_path)

# Confirmation
print("Connected to SQLite  database!")

Connected to SQLite  database!



# **STEP THREE**

Confirm the database contains the following four tables:

*   `customers`
*   `financials`
*   `sold_vehicles`
*   `trades`

By running the code block below.

In [2]:
# Returns all tables in database

query_tables = """SELECT name
                  FROM sqlite_schema
                  WHERE type='table'
                  ORDER BY name;"""

# Execute the query using pd.read_sql
df_tables = pd.read_sql(query_tables, conn)

print("These are the four tables in the database")
print(df_tables)

These are the four tables in the database
            name
0      customers
1     financials
2  sold_vehicles
3         trades


# **STEP FOUR**

## **Focus:** Total Revenue and Profit KPIs

To begin your analysis complete the prompts below that calculate the revenue and gross profit for the Elite Motors Dealership historical sales data.

> Remember, the definition for "revenue" is defined in the *Definitions* section of the **Luxury Dealership Business Use Case.pdf** resource included in the final project zip file.


## **`Prompts:`**
1.   Total revenue and total gross profit (total revenue - total cost) from all sold vehicles.
2.   Average gross profit per vehicle sold.
3.   Total number of vehicles sold yearly, for the last 5 years, by year in descending order.
4.   Total sales by vehicle make, excluding `Lexus`, with the 10 highest total sales at the top of the column. Confirm which vehicle make came in third highest and their total sales.
5.   Total number of sales by inventory type: new `N` and used `U`..



---

`Q1 PROMPT:` Total revenue and total gross profit (revenue - cost) from all sold vehicles.

In [ ]:
#Q1 QUERY

query_revenue_grossprofit ="""
SELECT SUM(sale_price) AS total_revenue, (SUM(sale_price) - SUM(total_cost)) AS total_gross_profit
FROM financials;
"""

df_revenue_grossprofit = pd.read_sql(query_revenue_grossprofit, conn)

print("Total Revenue and Total Gross Profit")
print(df_revenue_grossprofit)

Total Revenue and Total Gross Profit
   total_revenue  total_gross_profit
0   2.678816e+08        2.516746e+08


`Q1 ANSWER:`

Total Revenue is about $268 million.

Total Gross Profit is about $252 million.


`Q1 ANALYSIS:`
Given how close the total revenue is to the total gross profit, it seems like costs are minimized pretty well already. This leaves the other option left for increasing profit: increasing revenue itself. Perhaps the dealership should explore increasing prices in an attempt to increase revenue.


`Q2 PROMPT:` Average gross profit per vehicle sold.


In [ ]:
#Q2 QUERY

query_avg_profit = """
SELECT (SUM(sale_price) - SUM(total_cost))/COUNT(sale_price) AS avg_profit
FROM financials;
"""

df_avg_profit= pd.read_sql(query_avg_profit, conn)

print("Average Gross Profit: ")
print(df_avg_profit)

Average Gross Profit: 
     avg_profit
0  33228.758602


`Q2 ANSWER:`
Average gross profit per vehicle is about $33,229.

`Q2 ANALYSIS:`
The dealership could prioritize selling cars that provide a gross profit around this $33,229. For example, if there's a Lexus that usually sells for $36,000 but costs the dealership about $3000 total, that is the kind of car they should look to sell more of. Assuming a normal distribution where the average profit is also the most common value for profit, the easiest cars to sell in large quantities would be the ones that the dealership makes a ~$33,000 profit from.


`Q3 PROMPT:` Total number of vehicles sold yearly, for the last 5 years, by year in descending order.


In [243]:
#Q3 QUERY

query_total_sales = """
SELECT strftime('%Y', deal_date) AS year, COUNT(*) AS yearly_sales
FROM financials
GROUP BY year
ORDER BY year DESC
LIMIT 5;
"""

df_total_sales= pd.read_sql(query_total_sales, conn)

print("Total Sales by Year: ")
print(df_total_sales)

Total Sales by Year: 
   year  yearly_sales
0  2017           113
1  2016           585
2  2015           572
3  2014           598
4  2013           592


`Q3 ANSWER:`
The dealership has sold:

113 cars in 2017 (so far - data for 2017 only goes until March)

585 cars in 2016

572 cars in 2015

598 cars in 2014

592 cars in 2013


`Q3 ANALYSIS:`
The dealership should aim to follow recent trends of sales in the high 500s each year. Assuming it is currently the end of March in 2017, i.e. when this database ends, the dealership is on track to fall short of this goal for this year (113*4 = 452 cars by the end of the year). The dealership should push additional advertising to try and attract more customers for the rest of the year.


`Q4 PROMPT:` Total sales by vehicle make, excluding `Lexus`, with the 10 highest total sales at the top of the column. Confirm which vehicle make came in third highest and their total sales.


In [ ]:
#Q4 QUERY

query_make_sales = """
SELECT COUNT(*) AS total_sales, vehicle_make
FROM sold_vehicles
WHERE vehicle_make NOT IN ('lexus')
GROUP BY vehicle_make
ORDER BY total_sales DESC
LIMIT 10;
"""

df_make_sales= pd.read_sql(query_make_sales, conn)

print("Total Sales by Make: ")
print(df_make_sales)

Total Sales by Make: 
   total_sales   vehicle_make
0          210         toyota
1           96          honda
2           87  mercedes-benz
3           84           ford
4           78      chevrolet
5           76            bmw
6           73       cadillac
7           71           jeep
8           63         nissan
9           60       chrysler


`Q4 ANSWER:` Mercedes-Benz came in third place, with 87 total sales.

`Q4 ANALYSIS:`
After Lexus, the dealership seems to have the most success selling models from Toyota. They shoudl invest in more Toyota cars to see if that brings in more customers.


`Q5 PROMPT:` Total number of sales by inventory type: new `N` and used `U`.


In [238]:
#Q5 QUERY

query_new_used = """ 
SELECT COUNT(*) AS total_sales, inventory_type
FROM sold_vehicles
GROUP BY inventory_type
ORDER BY total_sales DESC;
"""

df_new_used= pd.read_sql(query_new_used, conn)

print("Total Sales by Inventory Type: ")
print(df_new_used)

Total Sales by Inventory Type: 
   total_sales inventory_type
0         3998              N
1         3576              U


`Q5 ANSWER:`
The dealership has sold 3998 new vehicles and 3576 used ones.

`Q5 ANALYSIS:`
Since the dealership sells more newer models, they should invest more in those newer cars rather than used ones.




---


# **Customer and Sales Performance:**

To continue your analysis complete these prompts below to identify top customers and sales performance by vehicle category.

## **`Prompts:`**

6.   Total vehicles sold for the top 5 customer cities in descending order by total sales. State which `city` is number 1 in total sales.
7.   Top ten postal codes with the highest overall sales as `total_sales` in descending order. The Elite Motors Dealership resides in the 46545 postal code. State the total sales for this postal code.
8.   State the most frequent and least frequent contract term types and their total count.

> **Note**, the observation of `1` in column `contract_term` means the customer paid for the vehicle with cash (1 time) and not through a loan.



`Q6 PROMPT:` Total vehicles sold for the top 5 customer cities in descending order by total sales. State which `city` is number 1 in total sales.


In [240]:
#Q6 QUERY

query_top_customers = """
SELECT COUNT(*) AS total_sales, city
FROM customers
GROUP BY city
ORDER BY total_sales DESC
LIMIT 5;
"""

df_top_customers= pd.read_sql(query_top_customers, conn)

print("Total Sales by City: ")
print(df_top_customers)

Total Sales by City: 
   total_sales          city
0         1512    South Bend
1          946       Granger
2          553       Elkhart
3          516     Mishawaka
4          251  Saint Joseph


`Q6 ANSWER:`
South Bend is the number 1 city in total sales, with 1512.

`Q6 ANALYSIS:`
The dealership should target its advertising towards South Bend and its surrounding area, as this is where they are able to attract the most customers.

`Q7 PROMPT:` Top ten postal codes with the highest overall sales as `total_sales` in descending order. The Elite Motors Dealership resides in the 46545 postal code. State the total sales for this postal code.


In [242]:
#Q7 QUERY

query_postalcodes = """
SELECT COUNT(*) AS total_sales, postal_code
FROM customers
GROUP BY postal_code
ORDER BY total_sales DESC
LIMIT 10;
"""

df_postalcodes= pd.read_sql(query_postalcodes, conn)

print("Total Sales by Postal Code: ")
print(df_postalcodes)

Total Sales by Postal Code: 
   total_sales postal_code
0          945       46530
1          568       46628
2          331       46514
3          324       46614
4          298       46545
5          253       49085
6          221       49120
7          215       46544
8          186       46637
9          160       46526


`Q7 ANSWER:`
The 46545 postal code has 298 total sales.

`Q7 ANALYSIS:`
By far the most popular postal code is 46530, which is Granger, Indiana. The dealership could invest more advertising into Granger to see if they can increase sales from this area even more. Conversely, this can also mean they attract plenty of customers from Granger already, and that other areas are an untapped market. Perhaps they can invest more into bringing customers from 46628, which is currently in second for total sales, to see if they can bring their sales numbers closer to that of 46530 and test if this theory is true.


`Q8 PROMPT:` Total sales as `term_total` by contract term type. State the most frequent and least frequent contract term types and their total count.

> Note, the observation of 1 in column `contract_term` means the customer paid for the vehicle with cash (1 time) and not through a loan.




In [244]:
#Q8 QUERY

query_term_total = """
SELECT COUNT(*) AS term_total, contract_term
FROM financials
GROUP BY contract_term
ORDER BY term_total DESC;
"""

df_term_total= pd.read_sql(query_term_total, conn)

print("Total Sales by Contract Term: ")
print(df_term_total)


Total Sales by Contract Term: 
    term_total  contract_term
0         4497              1
1         1297             60
2          792             36
3          615             48
4          114             72
5           72             24
6           37             66
7           31             39
8           21             42
9           17             75
10          16             51
11          15             63
12          12             27
13           9             84
14           8             77
15           7             54
16           6             33
17           3             45
18           3             30
19           2             69


`Q8 ANSWER:`

The most frequent contract term type is 1, with 4497 total sales.

The least frequent contract term type is 69, with 2 total sales.

`Q8 ANALYSIS:`
Seeing that the majority of customers (4497 out of 7574) pay using a one-time payment instead of through a loan, it's likely that most of the dealership's customers come from wealthy backgrounds that allow them to make these one-time payments. Not only can the dealership encourage this one-time payment system more, they can also target their advertising towards more affluent neighborhoods that have customers who are more likely to sign this kind of deal.

# **Trade-ins and Finance Analysis:**

To finalize your analysis complete these prompts below to identify the financial impact of trades on the overall gross profit for the dealership.

## **`Prompts:`**

9.   Total number of trade-in vehicles.

> **Note -** `Not Applicable` in column `trade_vin` means there was no vehicle traded-in at the time of the sale.

10.  Average profitability of sales transactions with a trade-in versus without a trade-in vehicle.

`Q9 PROMPT:` Total number of trade-in vehicles.

> **Note -** `Not Applicable` in column `trade_vin` means there was no vehicle traded-in at the time of the sale.






In [ ]:
#Q9 QUERY

query_trade_count = """
SELECT 
"""

df_trade_count = pd.read_sql(query_trade_count, conn)

print("Total Sales by Contract Term: ")
print(df_trade_count)

`Q9 ANSWER:`
*please state your answer to this query.*

`Q9 ANALYSIS:`
*please state the strategic suggestion or market opportunity you identified for the dealership*


`Q10 PROMPT:` Average profitability of sales transactions with a trade-in versus without a trade-in vehicle.

In [ ]:
#Q10 QUERY

query_profit =

`Q10 ANSWER:`
*please state your answer to this query.*

`Q10 ANALYSIS:`
*please state the strategic suggestion or market opportunity you identified for the dealership*
